<!DOCTYPE html>
<html lang="en">
<head>
<base target="_parent"/>
<link rel="stylesheet" media="all" href="/site_media/resources/bootstrap/css/bootstrap.min.css">
<link href="/site_media/resources/kendo/styles/kendo.common.min.css" rel="stylesheet">
<link href="/site_media/resources/less/kendo.tb.min.css" rel="stylesheet">
<link type="text/css" rel="stylesheet" href="/site_media/resources/styles/tb-styles.min.css">

<style type="text/css">
p, ul{font-family: "Lucida Sans Unicode", "Lucida Grande", Sans-Serif;}
.zebra {  background-color: #E1EFFA;}
.table_hover{    color: #009;    background-color: #6DBCEE;}
</style>

{% if autorefresh %}
<META HTTP-EQUIV="refresh" CONTENT="15">
{% endif %}

<style type="text/css">
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
</style>

</head>
<body>

<title>16S Metagenomics Analysis Report</title>

<style>
body {font-family: Arial;}
/* Style the tab */
.tab {
    overflow: hidden;
    border: 1px solid #ccc;
    background-color: #f1f1f1;
}

/* Style the buttons that are used to open the tab content */
.tab button {
    background-color: inherit;
    float: left;
    border: none;
    outline: none;
    cursor: pointer;
    padding: 14px 16px;
    transition: 0.3s;
}

/* Change background color of buttons on hover */
.tab button:hover {
    background-color: #ddd;
}

/* Create an active/current tablink class */
.tab button.active {
    background-color: #ccc;
}

/* Style the tab content */
.tabcontent {
    display: none;
    padding: 6px 12px;
    border: 1px solid #ccc;
    border-top: none;
}
</style>

<script>
function openTab(evt, tabName) {
    // Declare all variables
    var i, tabcontent, tablinks;

    // Get all elements with class="tabcontent" and hide them
    tabcontent = document.getElementsByClassName("tabcontent");
    for (i = 0; i < tabcontent.length; i++) {
        tabcontent[i].style.display = "none";
    }

    // Get all elements with class="tablinks" and remove the class "active"
    tablinks = document.getElementsByClassName("tablinks");
    for (i = 0; i < tablinks.length; i++) {
        tablinks[i].className = tablinks[i].className.replace(" active", "");
    }

    // Show the current tab, and add an "active" class to the button that opened the tab
    document.getElementById(tabName).style.display = "block";
    evt.currentTarget.className += " active";
}

// Get the element with id="defaultOpen" and click on it
document.getElementById("defaultOpen").click();

</script>

<h1><center>16S Metagenomics Analysis Report</center></h1>

<div class="container-fluid">


<h3> Overview </h3>
<div class="k-widget k-grid" >
<table class="table-striped">
          <tr class="d1"><td>16S Reference&nbsp;</td><td>:</td>              <td>{{genome_name}} </td></tr>
          <tr class="d1"><td>Library type&nbsp;</td><td>:</td>               <td>{{library_type}}</td></tr>
		  <tr class="d1"><td>Reads used for analysis&nbsp;</td><td>:</td>    <td>{{reads_num}}</td></tr>
          <tr><td>Download OTU matrix result&nbsp;</td><td>:</td> <td><a class="btn" href="all.unique.good.filter.unique.precluster.pick.pick.opti_mcc.shared" target="_blank">OTU matrix</a> </tr>
		  <tr><td>Download OTU taxonomy result&nbsp;</td><td>:</td> <td><a class="btn" href="all.unique.good.filter.unique.precluster.pick.pick.opti_mcc.0.03.cons.taxonomy" target="_blank">OTU taxonomy</a> </tr>
          <tr><td>Download OTU profile result&nbsp;</td><td>:</td> <td><a class="btn" href="all.OTU.summary.taxonomy" target="_blank">OTU profile table</a> </tr>
</table>
</div>
<br>

<h3> Meta16S workflow</h3>
<h4> <a href="Meta16S_workflow.png" target="_blank">Click here to download the figure below</a></h4>
<table class="block-rounded"><tr>
	<p style="text-align:center"><a href="Meta16S_workflow.png" title = "Meta16S_workflow"> <img style='width:679px;height:426px' src="Meta16S_workflow.png" alt="Meta16S workflow"> </a></p>
</table></tr>

<div class="spacer clear"></div>
<div class="spacer clear"></div>
<br>
<br>

<h3>Alpha diversity </h3>
<p>Alpha diversity is a measure of species richness or diversity within an individual sample. The simplest way of measuring it is the enumeration of OTUs
present in that sample, also called species richness (sobs). There are different diversity indices that capture also the structure of the community, including: 1)
chao - the Chao1 estimator; 2) shannon - the Shannon index; and 3) invsimpson - the inverse Simpson index.</p>
<h4> <a href="all.unique.good.filter.unique.precluster.pick.pick.opti_mcc.groups.summary"target="_blank">Click here to download the table below</a> </h4>
<div class="k-widget k-grid">
  <table class="table-striped">
    <thead class="k-grid-header">
      <tr>
      <th class="k-header"> <span class="help" title="label">label</span></th>
      <th class="k-header"> <span class="help" title="sample id">group</span></th>
      <th class="k-header"> <span class="help" title="sample coverage">coverage</span></th>
	  <th class="k-header"> <span class="help" title="number of sequence">nseqs</span></th>
      <th class="k-header"> <span class="help" title="number of observed OTUs">sobs</span></th>
      <th class="k-header"> <span class="help" title="chao">chao</span></th>
      <th class="k-header"> <span class="help" title="chao low confidence interval">chao_lci</span></th>
      <th class="k-header"> <span class="help" title="chao high confidence interval">chao_hci</span></th>
	  <th class="k-header"> <span class="help" title="shannon">shannon</span></th>
      <th class="k-header"> <span class="help" title="shannon low confidence interval">shannon_lci</span></th>
      <th class="k-header"> <span class="help" title="shannon high confidence interval">shannon_hci</span></th>
	  <th class="k-header"> <span class="help" title="invsimpson">invsimpson</span></th>
      <th class="k-header"> <span class="help" title="invsimpson low confidence interval">invsimpson_lci</span></th>
      <th class="k-header"> <span class="help" title="invsimpson high confidence interval">invsimpson_hci</span></th>
    </tr>
    </thead>
    {% for adiversity in adiversityData %}
       <tr>
          <td>{{adiversity.label}}</td>
		  
		  <td>{{adiversity.group}}</td>
		  
		  <td>{{adiversity.coverage}}</td>
		  
		  <td>{{adiversity.nseqs}}</td>
		  
		  <td>{{adiversity.sobs}}</td>
		  
		  <td>{{adiversity.chao}}</td>
		  
		  <td>{{adiversity.chao_lci}}</td>
		  
		  <td>{{adiversity.chao_hci}}</td>
		  
		  <td>{{adiversity.shannon}}</td>
		  
		  <td>{{adiversity.shannon_lci}}</td>
		  
		  <td>{{adiversity.shannon_hci}}</td>
		  
		  <td>{{adiversity.invsimpson}}</td>
		  
		  <td>{{adiversity.invsimpson_lci}}</td>
		  
		  <td>{{adiversity.invsimpson_hci}}</td>
        </tr>
    {% endfor %}
  </table>
</div>

<div class="spacer clear"></div>
<div class="spacer clear"></div>
<br>
<br>

<br>
<br>


<h3> Rarefaction Curve</h3>
<p>Rarefaction curves are useful for estimating the sufficiency of sequencing depth/effort per
sample. This is achieved by observing how the number of species (richness) change over
increasing number of reads. Rarefaction curves provide a way of comparing the richness observed in different samples. 
To plot curves, each sample is randomly subsampled without replacement at different intervals, and the average 
number of OTUs at each interval is plotted against the size of the sumsample. The point at which the number of OTUs
does not increase with further sampling is the point at which enough samples have been taken to accurately characterize 
the community. Different y-axis measurements can be plotted including: number of observed OTUs, Chao1 and Shannon index.
</p>

<!-- Tab links -->
<p>Select a measure:</p>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'sobs')">Number of observed OTUs</button>
  <button class="tablinks" onclick="openTab(event, 'shannon')">Shannon</button>
  <button class="tablinks" onclick="openTab(event, 'chao')">Chao1</button>
</div>

<!-- Tab content -->
<div id="sobs" class="tabcontent" style="display:block">
	<h4> <a href="RarefactionCurve.png" target="_blank">Click here to download the figure below</a></h4>
	<div class='center' style='width:550px;height:550px'>
		<td><a href="RarefactionCurve.png" title = "Rarefaction Curve"> <img style='width:500px;height:500px' src="RarefactionCurve.png" alt="Rarefaction Curve"> </a></td>
	</div>
</div>

<div id="shannon" class="tabcontent">
	<h4> <a href="RarefactionCurve_shannon.png" target="_blank">Click here to download the figure below</a></h4>
	<div class='center' style='width:550px;height:550px'>
		<td><a href="RarefactionCurve_shannon.png" title = "Rarefaction Measure: shannon"> <img style='width:500px;height:500px' src="RarefactionCurve_shannon.png" alt="Rarefaction Measure: shannon"> </a></td>
	</div>
</div>

<div id="chao" class="tabcontent">
	<h4> <a href="RarefactionCurve_chao.png" target="_blank">Click here to download the figure below</a></h4>
	<div class='center' style='width:550px;height:550px'>
		<td><a href="RarefactionCurve_chao.png" title = "Rarefaction Measure: chao1"> <img style='width:500px;height:500px' src="RarefactionCurve_chao.png" alt="Rarefaction Measure: chao1"> </a></td>
	</div>
</div>

<div class="spacer clear"></div>
<div class="spacer clear"></div>
<br>
<br>

<br>

<h3> Taxonomic binning</h3>
<p>For every taxonomic level, the relative abundances of all OTUs sharing the same taxonomy
are summed. This agglomeration of relative abundances is calculated also for sequences
of unknown taxonomic placement but otherwise sharing the same root. Stacked relative
abundance bar plots are produced for all samples at any given taxonomic level. This graphical 
representation may help to quickly visualize overall composition differences.</p>

<!-- Tab links -->
<p>Select a taxonomic level:</p>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'phylum')">Phylum</button>
  <button class="tablinks" onclick="openTab(event, 'class')">Class</button>
  <button class="tablinks" onclick="openTab(event, 'order')">Order</button>
  <button class="tablinks" onclick="openTab(event, 'family')">Family</button>
  <button class="tablinks" onclick="openTab(event, 'genus')">Genus</button>
  {% ifequal genome_name 'gg_13_8_99.fasta' %}
    <button class="tablinks" onclick="openTab(event, 'species')">Species</button>
  {% endifequal %}
</div>

<!-- Tab content -->
<div id="phylum" class="tabcontent">
	<tr>
		<td><a href="Taxonomic-Binning/1.Phyla.all.tab">Phylum</a></td>
	</tr>
	
	<div class='center' style='width:550px;height:550px'>
		<tr>
			<td><a href="Taxonomic-Binning/taxonomic-phyla.png"><img src="Taxonomic-Binning/taxonomic-phyla.png" alt="Phylum" width="500" height="500"></a></td>
		</tr>
	</div>
</div>

<div id="class" class="tabcontent">
	<tr>
		<td><a href="Taxonomic-Binning/2.Classes.all.tab">Class</a></td>
	</tr>
	
	<div class='center' style='width:550px;height:550px'>
		<tr>
			<td><a href="Taxonomic-Binning/taxonomic-class.png"><img src="Taxonomic-Binning/taxonomic-class.png" alt="Class" width="500" height="500"></a></td>
		</tr>
	</div>
</div>

<div id="order" class="tabcontent">
	<tr>
		<td><a href="Taxonomic-Binning/3.Orders.all.tab">Order</a></td>
	</tr>
	
	<div class='center' style='width:550px;height:550px'>
		<tr>
			<td><a href="Taxonomic-Binning/taxonomic-order.png"><img src="Taxonomic-Binning/taxonomic-order.png" alt="Order" width="500" height="500"></a></td>
		</tr>
	</div>
</div>

<div id="family" class="tabcontent">
	<tr>
		<td><a href="Taxonomic-Binning/4.Families.all.tab">Family</a></td>
	</tr>
	
	<div class='center' style='width:550px;height:550px'>
		<tr>
			<td><a href="Taxonomic-Binning/taxonomic-family.png"><img src="Taxonomic-Binning/taxonomic-family.png" alt="Family" width="500" height="500"></a></td>
		</tr>
	</div>
</div>


{% ifequal genome_name 'gg_13_8_99.fasta' %}
    <div id="genus" class="tabcontent">
	  <tr>
	    <td><a href="Taxonomic-Binning/5.Genera.all.tab">Genus</a></td>
	  </tr>
	
	  <div class='center' style='width:550px;height:550px'>
	    <tr>
		  <td><a href="Taxonomic-Binning/taxonomic-genus.png"><img src="Taxonomic-Binning/taxonomic-genus.png" alt="Genus" width="500" height="500"></a></td>
	    </tr>
	  </div>
    </div>

    <div id="species" class="tabcontent" style="display:block">
	  <tr>
	    <td><a href="Taxonomic-Binning/6.Species.all.tab">Species</a></td>
	  </tr>
	
	  <div class='center' style='width:550px;height:550px'>
	    <tr>
	      <td><a href="Taxonomic-Binning/taxonomic-species.png"><img src="Taxonomic-Binning/taxonomic-species.png" alt="Species" width="500" height="500"></a></td>
	    </tr>
	  </div>
    </div>

{% else %}
    <div id="genus" class="tabcontent" style="display:block">
	  <tr>
	    <td><a href="Taxonomic-Binning/5.Genera.all.tab">Genus</a></td>
	  </tr>
	
	  <div class='center' style='width:550px;height:550px'>
	    <tr>
		  <td><a href="Taxonomic-Binning/taxonomic-genus.png"><img src="Taxonomic-Binning/taxonomic-genus.png" alt="Genus" width="500" height="500"></a></td>
	    </tr>
	  </div>
    </div>
{% endifequal %}

<div class="spacer clear"></div>
<div class="spacer clear"></div>
<br>
<br>

<br>


<h3> Krona analysis</h3>
<h4> <a href="krona_analysis/krona.html" target="_blank">Click here to open the html file for Krona pie chart</a></h4>

<div class="spacer clear"></div>
<div class="spacer clear"></div>
<br>
<br>

<br>

<br>

<div class="spacer clear"></div>

{% if autorefresh %}
<br/><h3 style="text-align:center;color:red">*** Analysis is not complete ***</h3>
<a href="javascript:document.location.reload();" ONMOUSEOVER="window.status='Refresh'; return true">
<div style="text-align:center">Click here to refresh</div></a>
{% endif %}

</div>

<br/>
<div class="footer" style="width:100%"> <div class="container"> <div class="row"> <div class="span12">
  <img class="pull-left" src="/site_media/resources/img/ion-logo2.png" border="0">
  <div class="pull-right">
    <p>This is for research use only. It is not intended for any animal or human therapeutic or diagnostic use.</p>
    <p><span class="copyright">&copy; 2018 Thermo Fisher Scientific Inc. All Rights Reserved.</span>

    <a href="/licenses/terms-of-use.txt">Terms &amp; Conditions</a></p>
  </div>
</div> </div> </div> </div>
<br/>

</body> 
</html>
